### ISSUE TITLE
Cannot define permissions for global structs in C

### ISSUE THREAD

**Issue by ArmborstL** [utwente-fmt/vercors#1143]
The C example below has a global variable that is a struct pointer. It is impossible to define the permissions for that struct.
```c
struct accounts {
    int len;
};

struct accounts *accs;

/*@ requires Perm(accs, 1\2) ** accs != NULL ** Perm(*accs, 1\3);*/
void use_global() {
}

/*@ requires a != NULL ** Perm(a, 1\2) ** Perm(*a, 1\3);*/
void use_param(struct accounts *a) {
}
```
The `use_param` method shows how it is supposed to work: It is turned into the following Viper code (slightly simplified):
```
method useParam(tid: Int, a: Option[Pointer]) returns (res: Void)
  requires a != (none1(): Option[Pointer]) 
  requires acc(ptrDeref(optGet2(a)).ref, 1 / 2)
  requires acc(ptrDeref(optGet2(a)).ref.len, 1 / 3)
{}
```
Note that there is a level of indirection via the `ref` field. We get permission for that field via the term `Perm(a,1\2)` in the VerCors contract. Note that `a` is actually a local variable, so this term does not really make sense from the perspective of a VerCors user familiar with e.g. Java verification. Because it internally access the `ref` field of `a`, we actually first need to check that `a!=NULL`.

Now for `use_global`, the generated Viper code is like this:
```
method useGlobal(tid: Int) returns (res: Void)
  requires acc(globals().unknown, 1 / 2)
  requires globals().unknown != (none1(): Option[Pointer]) 
  requires acc(ptrDeref(optGet2(globals().unknown)).ref.len, 1 / 3)
{}
```
Here, the global variable `accs`, encoded as `globals().unknown`, is actually a heap variable. So the term `Perm(accs,1\2)` takes on the usual meaning, and only gives access to `accs` itself, not the field `ref`. As a result, there is no way to specify in the VerCors frontend that we want/need permission to its `ref` field. Consequently, the last clause, the permission for `len`, fails due to a lack of permission for `ref`. As a side note, the `NULL` check now has to occur _after_  the `Perm`, because it accesses the heap location.

I'm not sure what the solution for this should be. Personally, I find the fact that the local variable `a` has a `Perm(a,...)` confusing, and would prefer an encoding that does not use that. 
So maybe we should have something similar to `\pointer`, e.g. `\struct_pointer`. Then the precondition for `use_param` would look like `requires a != NULL ** \struct_pointer(a, 1\2) ** Perm(*a, 1\3)` and for `use_global` would be `requires Perm(accs, 1\2) ** accs != NULL ** \struct_pointer(accs, 1\2) ** Perm(*accs, 1\3)`. 
However, I'm not sure if this only occurs for structs, or also other pointers. Also, I'm not sure why the `ref` indirection is there in the first place. So maybe a different approach than the `\struct_pointer` is required.

**Comment by sakehl** [comment]
I discussed this wat Lukas yesterday.

So for my struct definition, and unpacking permissions, I never considered global variables. 

So normally, with non globals, this:
```
Perm(accs, 1\2)
```
leads to 
```
acc(ptrDeref(optGet2(a)).ref, 1 / 2)
```
But now you also need access towards the field of the global:
```
requires acc(globals().unknown, 1 / 2)
```
So we have no way to specify the permission towards the pointer.

So there are 3 permissions:
* Permission to the global variable
* Permission towards the pointer (`&accs[0]` or `accs`)
* Permission to the field `len` of the struct. (`accs->len`)


It seems we might need additional syntax to specify this with globals?

I'm not sure about `\struct_pointer` though. 

@pieter-bos any thoughts?

**Comment by pieter-bos** [comment]
Ah, this is just a mistake I made when designing heap variables, did it a bit too quickly I guess :)

My viewpoint is that in C all location arguments to `Perm` have to be a pointer value (so to evaluate `*a` you need permission to `a`, to evaluate `a[5]` you need permission to `a+5`, etc.) so it is wrong that `Perm(accs, _)` refers to the heap variable. Rather that should have already dereferenced the heap variable, and assert permission to the location referred to by the pointer value in the heap variable.

Tl;dr the notation for permission to a heap variable should be `Perm(&accs, read)` (in fact I think it's probably even wrong that it is a separate location type)

**Comment by sakehl** [comment]
Pointers of pointers, fun :p

But yeah it makes sense now, Perm(&accs, read) would be consistent with the other things in C.

**Comment by superaxander** [comment]
Fixed in #1345
